## ML Zoomcamp week 1 homework



#### Question 1

What's the version of Pandas that you installed?

You can get the version information using the __version__ field:

In [78]:
import pandas as pd
import numpy as np

In [2]:
pd.__version__

'2.1.0'

In [67]:
df = pd.read_csv('C:\\Users\\rlong\\downloads\\housing.csv')

In [22]:
df.shape

(20640, 10)

#### Question 2

How many columns are in the dataset?

    10
    6560
    10989
    20640


In [10]:
len(df.columns)

10

In [17]:
df.columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value', 'ocean_proximity'],
      dtype='object')

In [39]:
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

#### Question 3

Which columns in the dataset have missing values?

    total_rooms
    total_bedrooms
    both of the above
    no empty columns in the dataset


In [58]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

#### Question 4

How many unique values does the ocean_proximity column have?

    3
    5
    7
    9


In [29]:
df.ocean_proximity.unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

In [28]:
len(df.ocean_proximity.unique())

5

In [35]:
len(df[df['ocean_proximity'] == 'NEAR BAY'])

2290

#### Question 5

What's the average value of the median_house_value for the houses located near the bay?

    49433
    124805
    259212
    380440


In [52]:
df[df['ocean_proximity'] == 'NEAR BAY'].median_house_value.mean()

259212.31179039303

#### Question 6

    Calculate the average of total_bedrooms column in the dataset.
    Use the fillna method to fill the missing values in total_bedrooms with the mean value from the previous step.
    Now, calculate the average of total_bedrooms again.
    Has it changed?


In [59]:
df.total_bedrooms.mean()

537.8705525375618

In [68]:
df = df.fillna(df.total_bedrooms.mean())

In [69]:
df.total_bedrooms.isnull().sum()

0

In [70]:
df.total_bedrooms.mean()

537.8705525375617

#### Question 7

    Select all the options located on islands.
    Select only columns housing_median_age, total_rooms, total_bedrooms.
    Get the underlying NumPy array. Let's call it X.
    Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
    Compute the inverse of XTX.
    Create an array y with values [950, 1300, 800, 1000, 1300].
    Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
    What's the value of the last element of w?


In [112]:
df_island = df[df['ocean_proximity'] == 'ISLAND']
df_island

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
8314    -118.32     33.35                27.0       1675.0           521.0   
8315    -118.33     33.34                52.0       2359.0           591.0   
8316    -118.32     33.33                52.0       2127.0           512.0   
8317    -118.32     33.34                52.0        996.0           264.0   
8318    -118.48     33.43                29.0        716.0           214.0   

      population  households  median_income  median_house_value  \
8314       744.0       331.0         2.1579            450000.0   
8315      1100.0       431.0         2.8333            414700.0   
8316       733.0       288.0         3.3906            300000.0   
8317       341.0       160.0         2.7361            450000.0   
8318       422.0       173.0         2.6042            287500.0   

     ocean_proximity  
8314          ISLAND  
8315          ISLAND  
8316          ISLAND  
8317          ISLAND  
8318          ISLAND

In [123]:
df_small_island = df_island[['housing_median_age','total_rooms','total_bedrooms']]
df_small_island

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

In [124]:
X = df_small_island.to_numpy()
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [116]:
X.T

array([[  27.,   52.,   52.,   52.,   29.],
       [1675., 2359., 2127.,  996.,  716.],
       [ 521.,  591.,  512.,  264.,  214.]])

In [125]:
XTX = X.T.dot(X)
XTX

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [126]:
np.dot(X.T, X)

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [127]:
InvXTX = np.linalg.inv(XTX)
InvXTX

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [128]:
y = np.array([950, 1300, 800, 1000, 1300])

In [130]:
calc =  InvXTX.dot(X.T)
calc

array([[-0.00825608, -0.00653208, -0.00232159,  0.02565144,  0.01204934],
       [-0.00165852,  0.0011141 ,  0.00139656, -0.00103215, -0.00110698],
       [ 0.00754365, -0.00301964, -0.00455125,  0.00181685,  0.00329418]])

In [131]:
w = calc.dot(y)[2]
w

5.699229455065586